pennylane/ops/op_math/parametrized_hamiltonian.py

```diff
@@ -0,0 +1,172 @@
+
+    Args:
+        coeffs (Union[float, callable]): coefficients of the Hamiltonian expression, which may be constants or
+            parametrized functions. All functions passed as ``coeffs`` must accept the same parameters as arguments.
+        observables (Iterable[Observable]): observables in the Hamiltonian expression, of same length as coeffs
+
+    A ParametrizedHamiltonian is callable, and passing parameters to the ParametrizedHamiltonian will return an
+    Operator representing an instance of the Hamiltonian with the specified parameter values.
+
+    **Example:**
+
+    A ParametrizedHamiltonian can be created by passing a list of coefficients (scalars or functions), as well as
+    a list of corresponding observables. The functions must have identical signatures, though they may not all
+    use all the parameters.
+
+    >>> def f1(params, t): return np.sin(params[0]*t)
+    >>> def f2(params, t): return params[1] * np.cos(t)
+
+    The functions, along with scalar coefficients, can then be used to initialize a ParametrizedHamiltonian,
+    which will be split into a fixed and parametrized term. The fixed term is an Operator, while the parametrized
+    term must be initialized with concrete parameters to obtain an Operator.
+    >>> coeffs = [2, f1, f2]
+    >>> obs = [qml.PauliX(0)@qml.PauliX(1), qml.PauliY(0)@qml.PauliY(1), qml.PauliZ(0)@qml.PauliZ(1)]
+    >>> H = ParametrizedHamiltonian(coeffs, obs)
+    >>> H.H_fixed
+      2*(PauliX(wires=[0]) @ PauliX(wires=[1]))
+    >>> H.H_parametrized([2.5, 3.6], t)
+      (0.5984721441039564*(PauliY(wires=[0]) @ PauliY(wires=[1]))) + (1.9450883011253033*(PauliZ(wires=[0]) @ PauliZ(wires=[1])))
+
+    The resulting object can be passed parameters, and will return an Operator representing the
+    ParametrizedHamiltonian with the specified parameters:
+
+    >>> H([1.2, 2.3], 0.5)
+    (2*(PauliX(wires=[0]) @ PauliX(wires=[1]))) + ((0.5646424733950354*(PauliY(wires=[0]) @ PauliY(wires=[1]))) + (2.0184398923478573*(PauliZ(wires=[0]) @ PauliZ(wires=[1]))))
+
+    """
+
+    def __init__(self, coeffs, observables):
+
+        if qml.math.shape(coeffs)[0] != len(observables):
+            raise ValueError(
+                "Could not create valid Hamiltonian; "
+                "number of coefficients and operators does not match."
+            )
+
+        # ToDo: assuming this is how we want to do things, make this clear in the documentation
+        # since organizing into H_fixed and H_parametrized potentially moves around the order of the terms compared
+        # to how the terms were entered, sorting seems the least ambiguous organization for the wires
+        self.wires = Wires.all_wires([op.wires for op in observables], sort=True)
+
+        self._ops = list(observables)
+        self._coeffs = coeffs
+
+        self.H_coeffs_fixed = []
+        self.H_coeffs_parametrized = []
+        self.H_ops_fixed = []
+        self.H_ops_parametrized = []
```

Comment:
Out of curiosity, why do we draw a distinction between fixed and parametrized parts?

Reply:
Originally I was thinking of it in terms of a time-dependent and a drift term, and just separated them because I thought about them as separate things... Removing the distinction wouldn't break anything, we would just move what is happening in the `H_fixed` and `H_parametrized` methods to `__call__`. So its only purpose is if we are interested in being able to distinguish between the drift term and the rest. Possibly if we keep it this way, we should rename it to be `H_drift` instead of `H_fixed`.

Reply:
Thanks, yes I can see that making sense! `fixed` sounds fine to me (i.e., rather than `drift`)